twisted/protocols/test/test_tls.py

```diff
@@ -198,6 +215,104 @@ class NoFactory(object):
 
 
 
+def handshakingClientAndServer(clientGreetingData=None,
+                               clientAbortAfterHandshake=False):
+    """
+    Construct a client and server L{TLSMemoryBIOProtocol} connected by an IO
+    pump.
+
+    @param greetingData: The data which should be written in L{connectionMade}.
+    @type greetingData: L{bytes}
+
+    @return: 3-tuple of client, server, L{twisted.test.iosim.IOPump}
+    """
+    authCert, serverCert = certificatesForAuthorityAndServer()
+    @ClientFactory.forProtocol
+    @implementer(IHandshakeListener)
+    class Client(Protocol, object):
+        handshook = False
+        peerAfterHandshake = None
+
+        def connectionMade(self):
+            if clientGreetingData is not None:
+                self.transport.write(clientGreetingData)
+
+        def dataReceived(self, data):
+            self.applicationData.append(data)
+
+        def handshakeCompleted(self):
+            self.handshook = True
+            self.peerAfterHandshake = self.transport.getPeerCertificate()
+            if clientAbortAfterHandshake:
+                self.transport.abortConnection()
+
+    @ServerFactory.forProtocol
+    @implementer(IHandshakeListener)
+    class Server(Protocol, object):
+        handshaked = False
```

Comment:
So that they're clearly delineated in the tests and there's no chance that the objects being verified are backwards or duplicated.